Translate policy/limit-range.md in Korean (#19912)
kubernetes/website: content/ko/docs/concepts/policy/limit-range.md

---

> title: 범위 제한

Reviewer: ```suggestion
title: 리밋 레인지(Limit Range)
```

API 오브젝트이므로 외래어 표기법에 따라 표기해야합니다.

Author: 의견주신대로 타이틀 이후에 나오는 “LimitRange” 단어에 대해서 모두 “리밋레인지”로 변경해야겠지요?
이 부분은 추후 용어집에도 추가하는게 좋을 것 같습니다.

- 네임스페이스에서 리소스에 대한 요청과 제한 사이의 비율을 지정한다.
- 네임스페이스에서 컴퓨팅 리소스에 대한 기본 요청/제한을 설정하고 런타임에 있는 컨테이너에 자동으로 설정한다.

## LimitRange 활성화

많은 쿠버네티스 배포판에 LimitRange 지원이 기본적으로 활성화되어 있다. apiserver `--enable-admission-plugins=` 플래그의 인수 중 하나로 `LimitRanger` 어드미션 컨트롤러가 있는 경우 활성화된다.

해당 네임스페이스에 LimitRange 오브젝트가 있는 경우 특정 네임스페이스에 LimitRange가 지정된다.

LimitRange 오브젝트의 이름은 유효한 [DNS 서브도메인 이름](/ko/docs/concepts/overview/working-with-objects/names/#dns-서브도메인-이름)이어야한다.

### 범위 제한의 개요

- 관리자는 하나의 네임스페이스에 하나의 `LimitRange`를 만든다.
- 사용자는 네임스페이스에서 파드, 컨테이너 및 PersistentVolumeClaims과 같은 리소스를 생성한다.
- `LimitRanger` 어드미션 컨트롤러는 컴퓨팅 리소스 요청 사항을 설정하지 않은 모든 파드와 컨테이너에 대한 기본값과 제한을 지정하고 네임스페이스의 LimitRange에 정의된 리소스의 최소, 최대 및 비율을 초과하지 않도록 사용량을 추적한다.
- LimitRange 제약 조건을 위반하는 리소스(파드, 컨테이너, PersistentVolumeClaim)를 생성하거나 업데이트하는 경우 HTTP 상태 코드 `403 FORBIDDEN` 및 위반된 제약 조건을 설명하는 메시지와 함께 API 서버에 대한 요청이 실패한다.
- `cpu`, `memory`와 같은 컴퓨팅 리소스의 네임스페이스에서 LimitRange가 활성화된 경우 사용자는 해당 값에 대한 요청 또는 제한을 지정해야 한다. 그렇지 않으면 시스템에서 파드 생성이 거부될 수 있다.
- LimitRange 유효성 검사는 파드 실행 단계가 아닌 파드 어드미션 단계에서만 발생한다.

범위 제한을 사용하여 생성할 수 있는 정책의 예는 다음과 같다.

- 용량이 8GiB RAM과 16 코어인 2 노드 클러스터에서 네임스페이스의 파드를 제한하여 CPU의 최대 제한이 500m인 CPU 100m를 요청하고 메모리의 최대 제한이 600M인 메모리 200Mi를 요청하라.

> - 스펙에 CPU 및 메모리 요청없이 시작된 컨테이너에 대해 기본 CPU 제한 및 요청을 150m로, 메모리 기본 요청을 300Mi로 정의하라.

Reviewer 2: ```suggestion
- 16코어와 8 GiB RAM 용량이 있는 2 노드 클러스터에서 파드를 네임스페이스로 제한하여 CPU에 대해 100m 에서 최대 500m 으로 요청하고, 메모리는 200Mi 에서 최대 600Mi로 요청한다.
- 사양에 CPU 와 메모리 요구사양이 없는 상태에서 시작된 컨테이너의 경우 기본 CPU 제한과 요청을 150m 으로, 메모리 기본 요청을 300Mi로 정의한다.
```

조금 더 자연스럽게 변경해보았습니다.
만약 이를 받아들이시지 않으시더라도, 오타 수정과 띄어쓰기를 부탁드리겠습니다.

`8GiB` -> `8 GiB`, `600M` -> `600Mi`

4개의 컨테이너가 있는 파드가 먼저 생성된다. 파드 내의 각 컨테이너에는 특정 `spec.resource` 구성이 있다.
파드 내의 각 컨테이너는 `LimitRanger` 어드미션 컨트롤러에 의해 다르게 처리된다.

다음 kubectl 명령을 사용하여 네임스페이스 `limitrange-demo`를 생성한다.

```shell
kubectl create namespace limitrange-demo
```

kubectl 명령에서 네임스페이스 대상인 `limitrange-demo`를 빠트리지 않으려면 다음 명령으로 컨텍스트를 변경한다.

```shell
kubectl config set-context --current --namespace=limitrange-demo
```

다음은 LimitRange 오브젝트의 구성 파일이다.
{{< codenew file="admin/resource/limit-mem-cpu-container.yaml" >}}

이 오브젝트는 컨테이너에 적용할 최소 및 최대 CPU/메모리 제한, 기본 CPU/메모리 요청과 CPU/메모리 리소스에 대한 기본 제한을 정의한다.

다음 kubectl 명령을 사용하여 `limit-mem-cpu-per-container` LimitRange를 생성한다.

```shell
kubectl create -f https://k8s.io/examples/admin/resource/limit-mem-cpu-container.yaml
```

```shell
kubectl describe limitrange/limit-mem-cpu-per-container
```

```shell
Type        Resource  Min   Max   Default Request  Default Limit  Max Limit/Request Ratio
----        --------  ---   ---   ---------------  -------------  -----------------------
Container   cpu       100m  800m  110m             700m           -
Container   memory    99Mi  1Gi   111Mi            900Mi          -
```
다음은 4개의 컨테이너가 포함된 파드의 구성 파일로 LimitRange 기능을 보여준다.

> {{< codenew file="admin/resource/limit-range-pod-1.yaml" >}}

Reviewer 2: ```suggestion

다음은 4개의 컨테이너가 포함된 파드의 구성 파일로 LimitRange 기능을 보여준다.
{{< codenew file="admin/resource/limit-range-pod-1.yaml" >}}
```

원문에있는 공백이 사라졌습니다.

```

```json
{
  "limits": {
    "cpu": "500m",
    "memory": "200Mi"
  },
  "requests": {
    "cpu": "100m",
    "memory": "100Mi"
  }
}
```

- `busybox` 파드 내의 `busybox-cnt01` 컨테이너는 `requests.cpu=100m`와 `requests.memory=100Mi`로 정의됐다.
- `100m <= 500m <= 800m`, 컨테이너 CPU 제한(500m)은 승인된 CPU LimitRange 내에 있다.
- `99Mi <= 200Mi <= 1Gi`, 컨테이너 메모리 제한(200Mi)은 승인된 메모리 LimitRange 내에 있다.
- CPU/메모리에 대한 요청/제한 비율 검증이 없으므로 컨테이너가 유효하며 생성되었다.

### 유효한 CPU/메모리 요청은 있지만 제한이 없는 컨테이너 스펙

`busybox-cnt02`의 리소스 구성을 보자.

```shell
kubectl get po/busybox1 -o json | jq ".spec.containers[1].resources"
```

```json
{
  "limits": {
    "cpu": "700m",
    "memory": "900Mi"
  },
  "requests": {
    "cpu": "100m",
    "memory": "100Mi"
  }
}
```
- `busybox1` 파드 내의 `busybox-cnt02` 컨테이너는 `requests.cpu=100m`와 `requests.memory=100Mi`를 정의했지만 CPU와 메모리에 대한 제한은 없다.
- 컨테이너에 제한 섹션이 없다. `limit-mem-cpu-per-container` LimitRange 오브젝트에 정의된 기본 제한은 `limits.cpu=700mi` 및 `limits.memory=900Mi`로 이 컨테이너에 설정된다.
- `100m <= 700m <= 800m`, 컨테이너 CPU 제한(700m)이 승인된 CPU 제한 범위 내에 있다.
- `99Mi <= 900Mi <= 1Gi`, 컨테이너 메모리 제한(900Mi)이 승인된 메모리 제한 범위 내에 있다.
- 요청/제한 비율이 설정되지 않았으므로 컨테이너가 유효하며 생성되었다.

> ### 유효한 CPU/메모리 제한은 있지만 요청은 없는 컨테이너 스펙

Reviewer 2: ```suggestion
- 요청/제한 비율이 설정되지 않았으므로 컨테이너가 유효하며 생성되었다.

### 유효한 CPU/메모리 제한은 있지만 요청은 없는 컨테이너 스펙
```

원문에 있는 공백이 사라졌습니다.

`busybox-cnt03`의 리소스 구성을 보자.

```shell
kubectl get po/busybox1 -o json | jq ".spec.containers[2].resources"
```
```json
{
  "limits": {
    "cpu": "500m",
    "memory": "200Mi"
  },
  "requests": {
    "cpu": "500m",
    "memory": "200Mi"
  }
}
```

- `busybox1` 파드 내의 `busybox-cnt03` 컨테이너는 `limits.cpu=500m`와 `limits.memory=200Mi`를 정의했지만 CPU와 메모리에 대한 요청은 없다.
- 컨테이너에 요청 섹션이 정의되지 않았다. `limit-mem-cpu-per-container` LimitRange에 정의된 기본 요청은 제한 섹션을 채우는 데 사용되지 않지만 컨테이너에 의해 정의된 제한은 `limits.cpu=500m` 및 `limits.memory=200Mi`로 설정된다.
- `100m <= 500m <= 800m`, 컨테이너 CPU 제한(500m)은 승인된 CPU 제한 범위 내에 있다.
- `99Mi <= 200Mi <= 1Gi`, 컨테이너 메모리 제한(200Mi)은 승인된 메모리 제한 범위 내에 있다.
- 요청/제한 비율이 설정되지 않았으므로 컨테이너가 유효하며 생성되었다.

### CPU/메모리 요청/제한이 없는 컨테이너 스펙

`busybox-cnt04`의 리소스 구성을 보자.

```shell
kubectl get po/busybox1 -o json | jq ".spec.containers[3].resources"
```

```json
{
  "limits": {
    "cpu": "700m",
    "memory": "900Mi"
  },
  "requests": {
    "cpu": "110m",
    "memory": "111Mi"
  }
}
```

- `busybox1`의 `busybox-cnt04` 컨테이너는 제한이나 요청을 정의하지 않았다.

> - 컨테이너는 제한 섹션을 정의하지 않으며 `limit-mem-cpu-per-container` LimitRange에 정의된 기본 제한은 `limit.cpu=700m` 및 `limits.memory=900Mi`로 설정된다.

Reviewer 2: ```suggestion
- 컨테이너는 제한 섹션을 정의하지 않으며 `limit-mem-cpu-per-container` LimitRange에 정의된 기본 제한은
   `limit.cpu=700m` 및 `limits.memory=900Mi`로 설정된다.
```

원문에 맞추어 줄바꿈 했습니다.

```

`requests.storage`가 LimitRange의 Min 값보다 낮은 PVC를 만드는 동안 서버에서 발생하는 오류는 다음과 같다.

```shell
Error from server (Forbidden): error when creating "pvc-limit-lower.yaml": persistentvolumeclaims "pvc-limit-lower" is forbidden: minimum storage usage per PersistentVolumeClaim is 1Gi, but request is 500Mi.
```

`requests.storage`가 LimitRange의 Max 값보다 큰 경우에도 동일한 동작이 나타난다.

{{< codenew file="admin/resource/pvc-limit-greater.yaml" >}}

```shell
kubectl create -f https://k8s.io/examples/admin/resource/pvc-limit-greater.yaml
```

```shell
Error from server (Forbidden): error when creating "pvc-limit-greater.yaml": persistentvolumeclaims "pvc-limit-greater" is forbidden: maximum storage usage per PersistentVolumeClaim is 2Gi, but request is 5Gi.
```

## 제한/요청 비율

`LimitRangeSpec`에 `LimitRangeItem.maxLimitRequestRatio`가 지정되어 있으면 명명된 리소스는 제한을 요청으로 나눈 값이 열거된 값보다 작거나 같은 0이 아닌 값을 요청과 제한 모두 가져야 한다.

다음의 LimitRange는 메모리 제한이 네임스페이스의 모든 파드에 대한 메모리 요청 양의 최대 두 배가 되도록 한다.

{{< codenew file="admin/resource/limit-memory-ratio-pod.yaml" >}}

```shell
kubectl apply -f https://k8s.io/examples/admin/resource/limit-memory-ratio-pod.yaml
```

> 다음의 kubectl 명령으로 `limit-memory-ratio-pod` LimitRange의 정보를 나타낸다.

Reviewer 2: ```suggestion
다음의 kubectl 명령으로 <limit-memory-ratio-pod> LimitRange의 정보를 나타낸다.
```

원문에 맞추어 변경 부탁드립니다.

Author: 이 부분은 문서의 전체 내용 상 원문의 오타로 보입니다. 실제 출력시에 <>로 표시된 부분은 화면에 나타나지 않아 ``로 묶었습니다. 고쳐야할까요?

Reviewer 2: 기본적으로 원문(en)을 따라가는 것이 맞습니다.
다만, 잘못된 부분이라 판단되신다면 원문(en)에 수정 pr을 생성하셔서 원문에도 반영이 되도록 하시는 것도 하나의 방법입니다.

저 또한 번역을 진행하거나, 마일스톤 전환에 따른 문서 업데이트시 수정이 필요한 파일들에 대해서는 원문에 대한 수정 pr을 같이 올리기도 하고 있습니다.

Author: 의견 주신대로 원문에 대한 이슈와 PR을 올렸습니다. ( #20016 )
의견 감사합니다.

```shell
kubectl describe limitrange/limit-memory-ratio-pod
```

```shell
Name:       limit-memory-ratio-pod
Namespace:  limitrange-demo
Type        Resource  Min  Max  Default Request  Default Limit  Max Limit/Request Ratio
----        --------  ---  ---  ---------------  -------------  -----------------------
Pod         memory    -    -    -                -              2
```

> `requests.memory=100Mi` 및 `limits.memory=300Mi`로 파드를 생성한다.

Reviewer 2: ```suggestion

`requests.memory=100Mi` 및 `limits.memory=300Mi`로 파드를 생성한다.
```

원문에 없는 라인이 추가되었습니다.

{{< codenew file="admin/resource/limit-range-pod-3.yaml" >}}

```shell
kubectl apply -f https://k8s.io/examples/admin/resource/limit-range-pod-3.yaml
```

위 예에서 제한/요청 비율(`3`)이 `limit-memory-ratio-pod` LimitRange에 지정된 제한 비율(`2`)보다 커서 파드 생성에 실패했다.

```
Error from server (Forbidden): error when creating "limit-range-pod-3.yaml": pods "busybox3" is forbidden: memory max limit to request ratio per Pod is 2, but provided ratio is 3.000000.
```

## 정리

모든 리소스를 해제하려면 `limitrange-demo` 네임스페이스를 삭제한다.

```shell
kubectl delete ns limitrange-demo
```
다음 명령을 사용하여 컨텍스트를 `default` 네임스페이스로 변경한다.

```shell
kubectl config set-context --current --namespace=default
```

## 예제

- [네임스페이스별 컴퓨팅 리소스를 제한하는 방법에 대한 튜토리얼](/docs/tasks/administer-cluster/manage-resources/cpu-constraint-namespace/)을 참고하길 바란다.
- [스토리지 사용을 제한하는 방법](/docs/tasks/administer-cluster/limit-storage-consumption/#limitrange-to-limit-requests-for-storage)을 확인하라.

> - [네임스페이스별 쿼터에 대한 자세한 예](/docs/tasks/administer-cluster/quota-memory-cpu-namespace/)를 참고하길 바란다.

Reviewer 2: ```suggestion
- [네임스페이스별 컴퓨팅 리소스를 제한하는 방법에 대한 튜토리얼](/docs/tasks/administer-cluster/manage-resources/cpu-constraint-namespace/)을 본다.
- [스토리지 사용을 제한하는 방법](/docs/tasks/administer-cluster/limit-storage-consumption/#limitrange-to-limit-requests-for-storage)을 확인한다.
- [네임스페이스별 쿼터에 대한 자세한 예시](/docs/tasks/administer-cluster/quota-memory-cpu-namespace/)를 본다.
```

문체의 일관성을 위해 변경 부탁드립니다.

{{% /capture %}}

{{% capture whatsnext %}}

> 보다 자세한 내용은 [LimitRanger 설계 문서](https://git.k8s.io/community/contributors/design-proposals/resource-management/admission_control_limit_range.md)를 참고하길 바란다.

Reviewer 2: ```suggestion
보다 자세한 내용은 [LimitRanger 디자인 문서](https://git.k8s.io/community/contributors/design-proposals/resource-management/admission_control_limit_range.md)를 본다.
```

`design` 의 경우 다른 문서에서 `디자인` 으로 그대로 번역된 경우가 많아 선례를 따라 변경 부탁드립니다.
